### ISSUE TITLE
type x is not supported for the instanceof operation

### ISSUE THREAD

**Issue by ykoitzsch** [neow3j/neow3j#561]
I would like to check if the object I receive in the contract is of type X

```
public class X {
        public int x;
        public int y;
        public X() {}
    }
```

but using the instanceof operator ` if (data instanceof X) ` causes a compilation error:

```
The type 'io.neow3j.examples.contractdevelopment.contracts.frontier.p.Test$X' is not supported for the instanceof operation.
        at io.neow3j.compiler.converters.ObjectsConverter.handleInstanceOf(ObjectsConverter.java:105)
        at io.neow3j.compiler.converters.ObjectsConverter.convert(ObjectsConverter.java:89)
        at io.neow3j.compiler.Compiler.handleInsn(Compiler.java:404)
        at io.neow3j.compiler.NeoMethod.convert(NeoMethod.java:488)
        at io.neow3j.compiler.Compiler.compile(Compiler.java:271)
        at io.neow3j.compiler.Compiler.compile(Compiler.java:240)
```
using  ` if (data instanceof String) ` compiles. So maybe I am missing something here?

**Comment by csmuller** [comment]
The behavior of `instanceof` is not yet documented in neow3j.io. So, let me try to explain this here.  
Because all Java classes are simply represented by Arrays on the NeoVM, it is not possible to use the NeoVM instruction `ISTYPE` on regular Java objects like your class `X`. The `ISTYPE` instruction is the pendant to Java's `instanceof`, but it is not as powerful. It cannot tell two Arrays apart even if they were created for two different objects in the Java code.

In short, the neow3j compiler does not support what you want to do. The error message is intended and correct. 
The `instaceof` keyword can only be used with:
- String
- io.neow3j.devpack.ByteString
- Integer/int
- Long/long
- Byte/byte
- Short/short
- Character/char
- Boolean/boolean
- Byte[]/byte[]
- io.neow3j.devpack.Map
- io.neow3j.devpack.List
- io.neow3j.devpack.InteropInterface
- io.neow3j.devpack.Iterator.Struct

**Comment by csmuller** [comment]
I added this to the neow3j.io external docs.
Will close this issue now, since adding more type information to NeoVM stack items that represent NeowJava objects is not priority right now.

**Comment by ykoitzsch** [comment]
great, good to know and thank you :)
